### PR TITLE
Allow optional keywords in `create table` unique constraints

### PR DESCRIPTION
**Pull request by kayman-mk**
### Brief summary of the change made

Allows optional keywords and index name in `unique` constraints of `create table`.

```sql
CREATE TABLE a(
    a INT NOT NULL,
    UNIQUE KEY (a),
    UNIQUE INDEX (a),
    UNIQUE KEY idx_a(a),
    UNIQUE idx_b(a),
    UNIQUE (a),
);
```
- the index name is optional
- `key` and `index` keywords are optional

At least MySql and PostgreSql are using this syntax. I didn't found anything in SQL-99 but the implementation was already present in `dialect_mysql.py`

### Are there any other side effects of this change that we should be aware of?
No.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by jpy-git** [comment]
@kayman-mk I'm conscious that this PR is impacting the same area of the code as #2076 that we've just merged in. Can you update your fork and deal with any potential merge conflicts before I review?

**Comment by kayman-mk** [comment]
@jpy-git Branch updated. No conflicts found.

**Comment by jpy-git** [comment]
So I can see this implementation in the mysql docs: https://dev.mysql.com/doc/refman/8.0/en/create-table.html#:~:text=%5BCONSTRAINT%20%5Bsymbol%5D%5D%20UNIQUE%20%5BINDEX%20%7C%20KEY%5D

However reading the postgres docs the syntax for unique constraints seems to be different?: https://www.postgresql.org/docs/14/sql-createtable.html

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -2160,8 +2160,10 @@ class TableConstraintSegment(BaseSegment):
             "CONSTRAINT", Ref("ObjectReferenceSegment"), optional=True
         ),
         OneOf(
-            Sequence(  # UNIQUE ( column_name [, ... ] )
+            Sequence(  # UNIQUE [INDEX | KEY] [index_name] ( column_name [, ... ] )
                 "UNIQUE",
+                OneOf("INDEX", "KEY", optional=True),
+                Ref("ObjectReferenceSegment", optional=True),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_ansi.py:2166]: As far as I can see this is MySQL specific.

Can you copy this segment into the mysql dialect (changing the decorator to `@mysql_dialect.segment(replace=True)`) and remove these changes from ANSI.
i.e.
```sql
@mysql_dialect.segment(replace=True)
class TableConstraintSegment(BaseSegment):
    """A table constraint, e.g. for CREATE TABLE."""
    type = "table_constraint_segment"
    # Later add support for CHECK constraint, others?
    # e.g. CONSTRAINT constraint_1 PRIMARY KEY(column_1)
    match_grammar = Sequence(
        Sequence(  # [ CONSTRAINT <Constraint name> ]
            "CONSTRAINT", Ref("ObjectReferenceSegment"), optional=True
        ),
        OneOf(
            Sequence(  # UNIQUE [INDEX | KEY] [index_name] ( column_name [, ... ] )
                "UNIQUE",
                OneOf("INDEX", "KEY", optional=True),
                Ref("ObjectReferenceSegment", optional=True),
                Ref("BracketedColumnReferenceListGrammar"),
                # Later add support for index_parameters?
            ),
            Sequence(  # PRIMARY KEY ( column_name [, ... ] ) index_parameters
                Ref("PrimaryKeyGrammar"),
                # Columns making up PRIMARY KEY constraint
                Ref("BracketedColumnReferenceListGrammar"),
                # Later add support for index_parameters?
            ),
            Sequence(  # FOREIGN KEY ( column_name [, ... ] )
                # REFERENCES reftable [ ( refcolumn [, ... ] ) ]
                Ref("ForeignKeyGrammar"),
                # Local columns making up FOREIGN KEY constraint
                Ref("BracketedColumnReferenceListGrammar"),
                "REFERENCES",
                Ref("ColumnReferenceSegment"),
                # Foreign columns making up FOREIGN KEY constraint
                Ref("BracketedColumnReferenceListGrammar"),
                # Later add support for [MATCH FULL/PARTIAL/SIMPLE] ?
                AnyNumberOf(
                    # ON DELETE clause, e.g. ON DELETE NO ACTION
                    Sequence(
                        "ON",
                        "DELETE",
                        Ref("TableConstraintReferenceOptionGrammar"),
                    ),
                    # ON UPDATE clause, e.g. ON UPDATE SET NULL
                    Sequence(
                        "ON",
                        "UPDATE",
                        Ref("TableConstraintReferenceOptionGrammar"),
                    ),
                ),
            ),
        ),
    )
```

Also move your unit test to mysql

**Review by jpy-git — COMMENTED** [review]
@kayman-mk can you move this logic to mysql specific please?

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 1 above)
Comment (kayman-mk) [reply src/sqlfluff/dialects/dialect_ansi.py:2166]: Yeap. Let me check this. At least the MySql documentation says that it was introduced to make MySql compatible to other DBMS. But Postgres is definatly wrong.

It's on my list.

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 2]
```diff
@@ -2147,8 +2147,8 @@ class IndexColumnDefinitionSegment(BaseSegment):
     )
 
 
-@ansi_dialect.segment()
-class TableConstraintSegment(BaseSegment):
+@ansi_dialect.segmentTableConstraint()
+class Segment(BaseSegment):
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_ansi.py:2151]:
```suggestion
@ansi_dialect.segment()
class TableConstraintSegment(BaseSegment):
```
This shouldn't be changed

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 3]
```diff
@@ -230,6 +230,60 @@ class CreateTableStatementSegment(
         ],
     )
 
+@mysql_dialect.segmentTableConstraint(replace=True)
+class Segment(BaseSegment):
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_mysql.py:234]:
```suggestion
@mysql_dialect.segment(replace=True)
class TableConstraintSegment(BaseSegment):
```
as I suggested in the original review

**Review by jpy-git — CHANGES_REQUESTED** [review]

File: src/sqlfluff/dialects/dialect_mysql.py (hunk 3 above)
Comment (kayman-mk) [reply src/sqlfluff/dialects/dialect_mysql.py]: Right. No idea what I did there. Should be fixed with the latest commit.

**Review by jpy-git — APPROVED** [review]

**Comment by jpy-git** [comment]
@kayman-mk great work, thanks for adding this! 🥇

For some reason I don't have the usual option to update the branch for the latest changes in main 🤷 
image

Could you just pull in the latest changes from main and then we can get this merged in? Thanks 😄

**Comment by tunetheweb** [comment]
> For some reason I don't have the usual option to update the branch for the latest changes in main 🤷

When people open a PR there’s a tick box to allow maintainers to alter the branch. Most people leave it ticked to allow small changes and updates to be applied.

**Comment by jpy-git** [comment]
@tunetheweb ah makes sense thanks! 😄

**Pull request merged**